persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cluster/ClusterDistribution.scala

```diff
@@ -0,0 +1,128 @@
+import com.typesafe.config.Config
+
+import scala.concurrent.duration._
+
+/**
+ * Settings for cluster distribution.
+ *
+ * @param clusterShardingSettings The cluster sharding settings.
+ * @param ensureActiveInterval The interval at which entities are ensured to be active.
+ */
+case class ClusterDistributionSettings(
+  clusterShardingSettings: ClusterShardingSettings,
+  ensureActiveInterval:    FiniteDuration
+)
+
+object ClusterDistributionSettings {
+  def apply(system: ActorSystem): ClusterDistributionSettings = {
+    val clusterShardingSettings = ClusterShardingSettings(system)
+    val config = system.settings.config.getConfig("akka.cluster.distribution")
+    ClusterDistributionSettings(config, clusterShardingSettings)
+  }
+
+  def apply(config: Config, clusterShardingSettings: ClusterShardingSettings): ClusterDistributionSettings = {
+    val ensureActiveInterval = config.getDuration("ensure-active-interval", TimeUnit.MILLISECONDS).milliseconds
+    ClusterDistributionSettings(clusterShardingSettings, ensureActiveInterval)
+  }
+}
+
+object ClusterDistribution extends ExtensionId[ClusterDistribution] with ExtensionIdProvider {
+
+  override def createExtension(system: ExtendedActorSystem): ClusterDistribution =
+    new ClusterDistribution(system)
+
+  override def lookup = ClusterDistribution
+
+  /**
+   * Sent to each entity in a cluster distribution to ensure it's active.
+   */
+  case class EnsureActive(entityId: EntityId)
+}
+
+/**
+ * Distributes a static list of entities evenly over a cluster, ensuring that they are all continuously active.
+ *
+ * This uses cluster sharding underneath, using a scheduled tick sent to each entity as a mechanism to ensure they are
+ * active.
+ *
+ * Entities are cluster sharding entities, so they can discover their ID by inspecting their name. Additionally,
+ * entities should handle the [[com.lightbend.lagom.internal.persistence.cluster.ClusterDistribution.EnsureActive]]
+ * message, typically they can do nothing in response to it.
+ */
+class ClusterDistribution(system: ExtendedActorSystem) extends Extension {
+
+  /**
+   * Start a cluster distribution.
+   *
+   * @param typeName The name of the type of entity. This is used as the cluster sharding type name.
+   * @param entityProps The props for the entity actor.
+   * @param entityIds The entity ids to distribute over the cluster.
+   * @param settings The cluster distribution settings.
+   * @return the actor ref of the [[ShardRegion]] that is to be responsible for the shard
+   */
+  def start(
+    typeName:    String,
+    entityProps: Props,
+    entityIds:   Seq[EntityId],
+    settings:    ClusterDistributionSettings
+  ): ActorRef = {
+
+    val extractEntityId: ShardRegion.ExtractEntityId = {
+      case msg @ EnsureActive(entityId) => (entityId, msg)
+    }
+    val extractShardId: ShardRegion.ExtractShardId = {
+      case EnsureActive(entityId) => entityId
```

Comment:
The intention was that the entities are actually shards - afterall it's a static list and so small.  That said, we could have a shard function that turns an entity ID into a shard.

One question that I have, why isn't ExtractShardId a function of entity id -> shard id? Since the entity id has already been extracted out by ExtractEntityId?  My original intention was to allow the user to supply a custom extract entity id function which would mean they could then send their own messages to the entities as well, since I could combine the EnsureActive handling with orElse. But since extract shard id works off actual messages, and not entity IDs, they'd also have to supply a custom extract shard id function, but that would be a problem since the hash function of entity id to shard id must be consistent, we wouldn't be able to inject our own EnsureActive message handling to the extract shard id function, so the user would have to handle it in their own extract shard id.

I guess we can use the user defined extract entity id function to produce the extract shard id function, but it just seems odd to me the way it is, extract shard id could end up sending messages of the same entity id to different shards, is there ever a use case for doing that?